core/DCache.cpp

```diff
@@ -35,71 +66,246 @@ namespace olympia {
+        ILOG(mem_access_info_ptr << " in Lookup stage");
+        if (incoming_cache_refill_ == mem_access_info_ptr)
+        {
+            ILOG("Incoming cache refill " << mem_access_info_ptr);
+            return;
+        }
+
+        const bool hit = dataLookup_(mem_access_info_ptr);
+        ILOG(mem_access_info_ptr << " performing lookup " << hit);
+        if (hit)
+        {
+            mem_access_info_ptr->setCacheState(MemoryAccessInfo::CacheState::HIT);
+            out_lsu_lookup_ack_.send(mem_access_info_ptr);
+            return;
+        }
+
+        // Check MSHR Entries for address match
+        const auto & mshr_itb = mem_access_info_ptr->getMSHRInfoIterator();
+
+        if (!mshr_itb.isValid() && mshr_file_.numFree() == 0)
+        {
+            // Should be Nack but miss should work for now
+            mem_access_info_ptr->setCacheState(MemoryAccessInfo::CacheState::MISS);
+            out_lsu_lookup_ack_.send(mem_access_info_ptr);
+            return;
+        }
+
+        if (!mshr_itb.isValid())
+        {
+            //            mshrLookup_(mem_access_info_ptr);
+            if (!mem_access_info_ptr->getMSHRInfoIterator().isValid())
+            {
+                ILOG("Creating new MSHR Entry " << mem_access_info_ptr);
+                allocateMSHREntry_(mem_access_info_ptr);
             }
         }
+
+        replyLSU(mem_access_info_ptr);
+    }
+
+    void DCache::replyLSU(const MemoryAccessInfoPtr & mem_access_info_ptr)
+    {
+        const auto & mshr_it = mem_access_info_ptr->getMSHRInfoIterator();
+        const uint64_t block_addr = getBlockAddr(mem_access_info_ptr);
+        const bool data_arrived = (*mshr_it)->getDataArrived();
+        const bool is_store_inst = mem_access_info_ptr->getInstPtr()->isStoreInst();
+
+        // All ST are considered Hit
+        if (is_store_inst)
+        {
+            // Update Line fill buffer only if ST
+            ILOG("Write to Line fill buffer (ST), block address:0x" << std::hex << block_addr);
+            (*mshr_it)->setModified(true);
+            (*mshr_it)->setMemRequest(mem_access_info_ptr);
+            mem_access_info_ptr->setCacheState(MemoryAccessInfo::CacheState::HIT);
+            out_lsu_lookup_ack_.send(mem_access_info_ptr);
+            return;
+        }
+
+        if (data_arrived)
+        {
+            ILOG("Hit on Line fill buffer (LD), block address:0x" << std::hex << block_addr);
+            mem_access_info_ptr->setCacheState(MemoryAccessInfo::CacheState::HIT);
+            out_lsu_lookup_ack_.send(mem_access_info_ptr);
+            return;
+        }
+
+        // Enqueue Load in LMQ
+        ILOG("Load miss inst to LMQ; block address:0x" << std::hex << block_addr);
+        (*mshr_it)->setMemRequest(mem_access_info_ptr);
+        mem_access_info_ptr->setCacheState(MemoryAccessInfo::CacheState::MISS);
+        out_lsu_lookup_req_.send(mem_access_info_ptr);
```

Comment:
Move out_lsu_lookup_req_.send(...) out of both if statements. Have only one call to out port. This greatly simplifies debug.
Remove the return statements. Also simplifies debug.
Change second "if" to "else if"
Add "else" to enqueue load.
All in the name of debugging.